typegen/src/utils.rs

```diff
@@ -140,52 +155,68 @@ fn types_equal_inner(
         }
     }
 
-    let a_ty = types.resolve(a).expect("type a should exist in registry");
-    let b_ty = types.resolve(b).expect("type b should exist in registry");
-
     // Paths differ; types won't be equal then!
     if a_ty.path.segments != b_ty.path.segments {
         return false;
     }
 
-    // Names of type params don't line up, so different then!
-    if a_ty
-        .type_params
-        .iter()
-        .zip(&b_ty.type_params)
-        .any(|(a, b)| a.name != b.name)
-    {
-        return false;
-    }
+    let mut compare_fields = |a_params: &GenericsList,
+                              b_params: &GenericsList,
+                              a: &Field<PortableForm>,
+                              b: &Field<PortableForm>|
+     -> bool {
+        fn is_assoc(param: &Field<PortableForm>) -> bool {
+            param.type_name.as_ref().is_some_and(|x| x.contains("::"))
+        }
 
-    // We'll lazily extend our type params only if the shapes match.
-    let calc_params = || {
-        let a_params = a_parent_params.extend(&a_ty.type_params);
-        let b_params = b_parent_params.extend(&b_ty.type_params);
-        (a_params, b_params)
-    };
+        let equal_name = a.name == b.name;
+        let ty_indexes_deleted = a_params
+            .index_for_type_id(a.ty.id)
+            .zip(b_params.index_for_type_id(b.ty.id))
+            .is_none();
 
-    // Capture a few variables to avoid some repetition later when we recurse.
-    let mut types_equal_recurse =
-        |a: u32, a_params: &GenericsList, b: u32, b_params: &GenericsList| -> bool {
-            types_equal_inner(a, a_params, a_visited, b, b_params, b_visited, types)
+        if !equal_name {
+            return false;
         };
 
+        match (&a.type_name, &b.type_name) {
+            // both generic but params were skipped
+            // removing this breaks utils::tests::recursive_data
+            (Some(_), Some(_)) if ty_indexes_deleted => {
+                types_equal_recurse(a.ty.id, a_params, b.ty.id, b_params)
+            }
+            (Some(a_type_name), Some(b_type_name)) => {
+                let type_names_present = match (
+                    a_params.index_for_type_name(a_type_name),
+                    b_params.index_for_type_name(b_type_name),
+                ) {
+                    // generic param idxs are the same
+                    (Some(x), Some(z)) => x == z,
+                    // no indexed present so we compare just the type names
+                    (None, None) => a_type_name == b_type_name,
+                    _ => false,
+                };
+
+                !is_assoc(a) && !is_assoc(b) && type_names_present
```

Comment:
or, since we have the type name strings to hand here: `!a_type_name.contains("::") && !b_type_name.contains("::") && type_names_are_equal`?

What would break if this test for "::" was removed?

Reply:
simplified the whole section completetely after some debugging